File: app/lib/browser/tab_manager.dart
import 'dart:collection';
import 'package:flutter/foundation.dart';
import 'browser_tab.dart';

/// Sequential tab model (mirrors the Java TabManager's invariant):
/// new tabs insert immediately after the active tab; closing the active tab
/// activates its sequential neighbor; only an explicit tap jumps arbitrarily.
class TabManager extends ChangeNotifier {
  final List<BrowserTab> _tabs = [];
  int _activeIndex = -1;
  int _nextId = 0;

  List<BrowserTab> get tabs => UnmodifiableListView(_tabs);
  BrowserTab? get active => _activeIndex >= 0 && _activeIndex < _tabs.length ? _tabs[_activeIndex] : null;
  int get length => _tabs.length;
  int get activeIndex => _activeIndex;

  /// Create a tab, inserted immediately after the active tab.
  BrowserTab newTab({bool isPrivate = false, String url = ''}) {
    final tab = BrowserTab(_nextId++, isPrivate: isPrivate, url: url);
    final ai = _activeIndex;
    if (ai >= 0 && ai < _tabs.length - 1) {
      _tabs.insert(ai + 1, tab);
      _activeIndex = ai + 1;
    } else {
      _tabs.add(tab);
      _activeIndex = _tabs.length - 1;
    }
    notifyListeners();
    return tab;
  }

  /// Explicitly switch to a tab (the ONLY non-adjacent active-pointer change).
  void select(int index) {
    if (index >= 0 && index < _tabs.length) {
      _activeIndex = index;
      notifyListeners();
    }
  }

  /// Explicitly switch to a tab by its stable id.
  void selectById(int id) {
    final i = _tabs.indexWhere((t) => t.id == id);
    if (i >= 0) select(i);
  }

  /// Close a tab by its stable id (grid cards hold ids, not positions).
  void closeById(int id) {
    final i = _tabs.indexWhere((t) => t.id == id);
    if (i >= 0) close(i);
  }

  void close(int index) {
    if (index < 0 || index >= _tabs.length) return;
    final wasActive = index == _activeIndex;
    final tab = _tabs.removeAt(index);
    tab.dispose();
    if (_tabs.isEmpty) {
      _activeIndex = -1;
    } else if (wasActive) {
      _activeIndex = index > 0 ? index - 1 : 0; // sequential neighbor
    } else if (_activeIndex > index) {
      _activeIndex--;
    }
    notifyListeners();
  }

  DateTime? _lastApply;
  bool _applyInProgress = false;
  final Set<String> _pendingChangedKeys = {};

  /// Push the current user settings (JavaScript, parameter stripping) into
  /// every live engine, so a Settings change reaches pages already open.
  /// P0 fix: debounced and parallel, not sequential await per tab.
  Future<void> applySettingsToAll({bool force = false}) async {
    if (_applyInProgress && !force) return;
    final now = DateTime.now();
    if (!force && _lastApply != null && now.difference(_lastApply!).inMilliseconds < 500) {
      // Root fix: accumulate changed keys during debounce window instead of dropping
      _pendingChangedKeys.addAll(_pendingChangedKeys); // keep existing
      return;
    }
    _lastApply = now;
    _applyInProgress = true;
    final keysToApply = Set<String>.from(_pendingChangedKeys)..clear();
    try {
      await Future.wait(_tabs.map((tab) => tab.engine.applySettings().catchError((_) {})));
    } finally {
      _applyInProgress = false;
    }
  }

  /// Only call when settings that affect WebView actually changed (js, blocking, etc)
  Future<void> applySettingsIfNeeded({required Set<String> changedKeys}) async {
    const webViewKeys = {'js', 'blocking', 'paramStripping', 'resourcePolicy', 'profile'};
    final relevant = changedKeys.intersection(webViewKeys);
    if (relevant.isEmpty) return;
    _pendingChangedKeys.addAll(relevant);
    await applySettingsToAll(force: true);
  }

  /// Remove private tab models and await their native retained-page teardown.
  /// Clear Data calls this before profile deletion; normal tabs stay open and
  /// reload against the now-empty default stores.
  Future<void> closePrivateTabs() async {
    final closing = <BrowserTab>[];
    for (var i = _tabs.length - 1; i >= 0; i--) {
      if (_tabs[i].isPrivate) {
        closing.add(_tabs[i]);
        close(i);
      }
    }
    await Future.wait(closing.map((tab) => tab.releaseNativeOwnership()));
  }

  void closeAll() {
    for (final t in _tabs) {
      t.dispose();
    }
    _tabs.clear();
    _activeIndex = -1;
    notifyListeners();
  }
}

File: android/app/src/main/java/com/mrnobody/browser/ui/TabManager.java
package com.mrnobody.browser.ui;

import java.util.ArrayList;
import java.util.List;

/**
 * Owns the list of tabs and the notion of an active tab.
 *
 * Tab order == navigation order. A new tab is inserted immediately after the
 * active tab (not appended to the far end), and closing the active tab activates
 * its sequential neighbor (the tab before it, else the one after). Only an
 * explicit {@link #setActive} jumps to an arbitrary tab — nothing else reorders.
 */
public final class TabManager {

    private final List<Tab> tabs = new ArrayList<>();
    private int activeIndex = -1;
    private int nextId = 0;

    /** Create a new tab, inserted immediately after the active tab. */
    public Tab newTab(boolean isPrivate) {
        return insert(nextId++, isPrivate, "", "");
    }

    /** Create a restored tab with a specific id/url/title (state memory restore). */
    public Tab restoreTab(int id, boolean isPrivate, String url, String title) {
        if (id >= nextId) nextId = id + 1;
        Tab t = insert(id, isPrivate, url, title);
        // restore keeps order; do NOT auto-activate — caller sets the active tab.
        return t;
    }

    private Tab insert(int id, boolean isPrivate, String url, String title) {
        Tab t = new Tab(id, isPrivate);
        t.setUrl(url);
        t.setTitle(title);
        int ai = activeIndex;
        if (ai >= 0 && ai < tabs.size() - 1) {
            tabs.add(ai + 1, t);          // immediately after the active tab
            activeIndex = ai + 1;
        } else {
            tabs.add(t);                  // append (no active tab, or active is last)
            activeIndex = tabs.size() - 1;
        }
        return t;
    }

    public Tab getActive() {
        return activeIndex >= 0 && activeIndex < tabs.size() ? tabs.get(activeIndex) : null;
    }

    public Tab get(int index) {
        return (index >= 0 && index < tabs.size()) ? tabs.get(index) : null;
    }

    /** Find a tab by id (for restore + deep-link routing). */
    public Tab findById(int id) {
        for (Tab t : tabs) if (t.id() == id) return t;
        return null;
    }

    /** Explicitly jump to a tab. This is the ONLY way the active pointer moves non-adjacently. */
    public void setActive(int index) {
        if (index >= 0 && index < tabs.size()) activeIndex = index;
    }

    public void setActiveById(int id) {
        int i = indexOf(findById(id));
        if (i >= 0) activeIndex = i;
    }

    public List<Tab> all() {
        return tabs;
    }

    public int size() {
        return tabs.size();
    }

    public int indexOf(Tab tab) {
        return tabs.indexOf(tab);
    }

    public int nextId() {
        return nextId;
    }

    /** Bump the id counter past any restored ids. */
    public void setNextId(int id) {
        if (id > nextId) nextId = id;
    }

    public void close(int index) {
        if (index < 0 || index >= tabs.size()) return;
        boolean wasActive = index == activeIndex;
        Tab t = tabs.remove(index);
        t.destroy();
        if (tabs.isEmpty()) {
            activeIndex = -1;
        } else if (wasActive) {
            // activate the sequential neighbor: the tab before, else the one after
            activeIndex = Math.max(0, index - 1);
        } else if (activeIndex > index) {
            activeIndex--;
        }
    }

    public void closeById(int id) {
        Tab t = findById(id);
        if (t != null) close(indexOf(t));
    }

    public void closeAll() {
        for (Tab t : tabs) t.destroy();
        tabs.clear();
        activeIndex = -1;
    }
}
